### PR TITLE
implementing the Checkout component

### PR DESCRIPTION
**Pull request by julien51**
# Description

This adds the component for checkout which can be used by multiple tools.
There are a couple things we should improve, including wha happens whe the user clicks on "Go to Content".

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [X] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [X] My code follows the style guidelines of this project, including naming conventions
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] If my code adds or changes components, I have written corresponding stories with Storybook
- [X] I have performed a self-review of my own code
- [X] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: paywall/src/components/checkout/Checkout.tsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+export const Checkout = ({
+  locks,
+  config,
+  account,
+  // keys,
+  purchase,
+}: Props) => {
+  const hasValidKey = Object.keys(locks).reduce((hasValidKey, address) => {
```
Comment (cellog) [inline paywall/src/components/checkout/Checkout.tsx]: this can become:

```js
const hasValidKey = Object.values(locks).filter(lock => lock.key.status === 'valid').length
```

File: paywall/src/unlockTypes.ts [hunk 2]
```diff
@@ -51,3 +51,61 @@ export interface Network {
 export interface ChainExplorerURLBuilders {
   [site: string]: (address: string) => string
 }
+
+export interface Lock {
+  address: string
+  transaction: string
+}
```
Comment (cellog) [inline paywall/src/unlockTypes.ts]: can add `key: Key`

File: paywall/src/unlockTypes.ts [hunk 3]
```diff
@@ -51,3 +51,61 @@ export interface Network {
...
+
+export interface Key {
+  lock: string
+  id: string
+  owner: string
+  status: string
+  expiration: number
+  transactions?: Transaction[]
```
Comment (cellog) [inline paywall/src/unlockTypes.ts]: this will always be set, you can remove the `?`

Comment (cellog) [reply paywall/src/unlockTypes.ts]: also, add `confirmations: number` (this too will always be set)

File: paywall/src/unlockTypes.ts [hunk 4]
```diff
@@ -51,3 +51,61 @@ export interface Network {
...
+  locks: Locks
+}
+
+export enum KeyStatus {
+  NONE = 'none',
+  CONFIRMING = 'confirming',
+  CONFIRMED = 'confirmed',
+  EXPIRED = 'expired',
+  VALID = 'valid',
+}
```
Comment (cellog) [inline paywall/src/unlockTypes.ts]: let's add 

```ts
  SUBMITTED = 'submitted',
  PENDING = 'pending',
  FAILED = 'failed',
```

File: paywall/src/unlockTypes.ts [hunk 5]
```diff
@@ -51,3 +51,61 @@ export interface Network {
...
+
+export interface UILocks {
+  [address: string]: UILock
+}
+
+export interface UIKey {
+  expiration: number
+  transactions: Transaction[]
+  status: string
```
Comment (cellog) [inline paywall/src/unlockTypes.ts]: can add `confirmations: number` here too if that makes sense

File: paywall/src/components/checkout/Checkout.tsx (hunk 1 above)
Comment (cellog) [reply paywall/src/components/checkout/Checkout.tsx]: https://stackoverflow.com/questions/42966362/how-to-use-object-values-with-typescript/42967397#comment90677819_42967397 may help us here

Comment (julien51) [reply paywall/src/components/checkout/Checkout.tsx]: In the end, is that cleaner than using `Object.keys`?

Comment (cellog) [reply paywall/src/components/checkout/Checkout.tsx]: no. let me rephrase using Object.keys:

```js
const hasValidKey = Object.keys(locks).reduce((isValid, address) => locks[address].key.status === 'valid', false)
```

Don't do any calculation

Comment (julien51) [reply paywall/src/components/checkout/Checkout.tsx]: Oh... that's what you meant, using the key status rather than the expiration! Sorry! Yeah, I'll do that.

**Comment by julien51** [comment]
image


image


image


image

image

image

image

**Comment by julien51** [comment]
Ok @cellog added screenshots and I think this is ready for review now!

**Comment by cellog** [comment]
one question: do we want to dim the other locks if a purchase is in progress?

**Comment by julien51** [comment]
I thought about this and I am not sure... I am actually not sure we should dim them when there is a valid key.

File: paywall/src/unlockTypes.ts (hunk 5 above)
Comment (cellog) [reply paywall/src/unlockTypes.ts]: key will also have a lock member, which is the lock address. Since it's not needed, you can include it for completeness or not
